docs/rfc/rfc-004-e2e-framework.rst

```diff
@@ -0,0 +1,212 @@
+of test cases to support parallelism. The goal is to extend this pattern and
+avoid hardcoding the matrix of test cases in the generator code.  Once the
+testnet configuration generation behavior is configurable at runtime,
+developers may be able to use the e2e framework to validate changes before
+landing changes that break e2e tests a day later.
+
+In addition to the autogenerated suite, it might make sense to maintain a
+small collection of hand-crafted cases that exercise configurations of
+concern, to run as part of the nightly (or less frequent) loop.
+
+Implementation Plan Structure
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As a development team, we should determine the features should impact the e2e
+testing early in the development cycle, and if we intend to modify the e2e
+tests to exercise a feature, we should identify this early and begin the
+integration process as early as possible.
+
+To facilitate this, we should addopt a practice whereby we exercise specific
+features that are currently under development more rigorously in the e2e
+suite, and then as development stabilizes we can reduce the number or weight
+of these features in the suite.
+
+There are essentially two end to end tests: the suite of (as of 0.35) 64
+different test networks, and the hand crafted `ci.toml` test case. The
+generated test cases help provide
+
+Reduce Cycle Time
+~~~~~~~~~~~~~~~~~
+
+One of the barriers to leveraging the e2e framework, and one of the challenges
+in debugging failures, is the cycle time of running a single test iteration is
+quite high: 5 minutes to build the docker image, plus the time to run the test
+or tests.
+
+There are a number of improvements and enhancements that can reduce the cycle
+time in practice:
+
+- reduce the amount of time required to build the docker image used in these
+  tests. Without the dependency on CGo, the tendermint binaries could be
+  (cross) compiled outside of the docker container and then injected into
+  them, which would take better advantage of docker's native caching,
+  although, without the dependency on CGo there would be no hard requirement
+  for the e2e tests to use docker.
+
+- support test parallelism. Because of the way the testnets are orchestrated
+  a single system can really only run one network at a time. For executions
+  (local or remote) with more resources, there's no reason to run a few
+  networks in parallel to reduce the feedback time.
+
+- prune testnet configurations that are unlikely to provide good signal, to
+  shorten the time to feedback.
+
+- apply some kind of tiered approach to test execution, to improve the
+  legibility of the test result. For example order tests by the dependency of
+  their features, or run test networks without perturbations before running
+  that configuration with perturbations, to be able to isolate the impact of
+  specific features.
+
+- orchestrate the test harness directly from go test rather than via a special
```

Comment:
This would be nice, but might turn into a Whole Thing pretty quickly. I guess we aren't really doing anything that complicated with docker-compose, though, so maybe it's not too bad.

Reply:
I think there are two axis of orchestration, here: 

- managing the process (docker-compose) 
- running all of the cases/networks (runner, run-multiple.sh, etc.)

They're both kind of orthogonal, but I think moving run-multiple.sh into go means that we can run two networks at the same time (to avoid port conflicts,) and have better control over data collection, while getting rid of docker compose mostly gets us out of build system dependencies and gives us more introspection into the processes themselves (probably?) 

Putting things in go test, from my perspective is probably more about developer interface and simplifying CI configuration.

Reply:
I understand that it has some rough edges, but I find some features of docker reasonably nice for tasks like this (this doc is possibly not the place to totally hash that out). 

Docker provides a way of grouping process metrics via cgroups that could be useful for process introspection when running multiple networks simultaneously. I also think we could solve the conflicting port problems. It looks like each testnet uses a separate docker network name which will generate a different virtual interface/bridge where the containers in separate testnets should be able to allocate their own ports. I think right now the problem may be that the e2e suite is just removing all other docker networks on start so only one can be run at a time, although I could be wrong on that.

Reply:
yeah, I think the decisions (which I don't think we should make right now) are: 

- how to orchestrate processes
- what's the interface for tests
- how to orchestrate test networks

all of which or orthogonal (mostly), and I think there are merits to using docker (and also not)

Reply:
I basically agree. It's true that configuring networks locally without some kind of virtualization is fraught. For better or worse, Docker is probably the easiest way to manage that. (And once we're in for that, the process/container bits are mostly free, though for what we do it probably _would_ be easier without docker-compose in the loop)

Reply:
I think my main angle that's different from y'all is that I'm not sure (or not convinced?) we need to have virtualization for what we're trying to do? our only requirement is port separation...

Reply:
It doesn't really require machine virtualization—thats just an easy way to provision a network with a NAT so you can use multiple IPs. Ideally we should at least _have_ the option to run nodes on different IPs (and we should probably be testing IPv6 harder too). 

I guess we could install and set up wireguard, but that's probably no easier than what we're doing now to coordinate. I don't think we gain all that much from the containerization otherwise, since we're running Go binaries.

Anyway, though, agreed with what y'all said above, that we don't have to sort it all out right here/now. 🙂